Add converter for `typedef` property

hyperupcall opened the pull request:
<!--
👋 Hi, thanks for sending a PR to tslint-to-eslint-config! 💖
Please fill out all fields below to ensure your PR is reviewed quickly.
-->
Hoi! ^w^ :wave:

## PR Checklist

-   [x] Fixes #937
-   [x] That issue was marked as [`status: accepting prs`](https://github.com/typescript-eslint/tslint-to-eslint-config/labels/status%3A%20accepting%20prs)

## Overview

The [typedef](https://palantir.github.io/tslint/rules/typedef/) property of TSLint is now considered during the conversion. All config options except `call-signature` and `arrow-call-signature` are accounted for. Options unaccounted for are mentioned in the `notices` array, when applicable

The linked issue and the [ESLint typedef docs](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/typedef.md) mentions that [explicit-function-return-type](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/explicit-function-return-type.md) and [explicit-module-boundary-types](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/explicit-module-boundary-types.md) should be used when converting `call-signature` and `arrow-call-signature`. I did that in a separate branch [here](https://github.com/hyperupcall/tslint-to-eslint-config/commit/b99f54eaa8a4bc1e1fe0729b6b80d5069f77572d), but I wasn't 100% confident that my conversion was correct, so I left it out for now. If it is close to what is considered acceptable, maybe I can add it to this PR?

I'll also add that I had to install `@swc/core` to get things to run, despite [this](65f9090345b31f22371fcdffea22be5b833cbe2dc) mentioning its not required - not sure what's up with that. Also the Docs say Node 12 works, but 14 was the minimum LTS that worked for me (Jest threw `SyntaxError: Unexpected token '.'` when it encountered optional chaining)

hyperupcall committed:
feat: Add converter for `typedef` rule (#937)

JoshuaKGoldberg reviewed (changes requested):
Awesome, looks good so far. Glad to see this rule finally getting taken care of! 👏

Tests are failing ([ci link](https://app.circleci.com/pipelines/github/typescript-eslint/tslint-to-eslint-config/2825/workflows/a2ef8400-b46d-49c9-a9f0-db9f1da268e6/jobs/14581)) because a lack of coverage on lines `11,15,19,27,31`. You'll want to add >=1 test case that covers them.

JoshuaKGoldberg commented:
> install @swc/core to get things to run

Huh, sorry about that - I'll send a separate PR to add it in.

JoshuaKGoldberg commented:
>  Node 12 works, but 14 was the minimum LTS that worked for me (Jest threw SyntaxError: Unexpected token '.' when it encountered optional chaining)

Nice, following up on that too: https://github.com/typescript-eslint/tslint-to-eslint-config/pull/1374

hyperupcall commented on src/converters/lintConfigs/rules/ruleConverters/typedef.ts:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    return {
+        ...(tslintRule.ruleArguments.some(
```
Good idea :+1: I'll make those changes

hyperupcall commented on src/converters/lintConfigs/rules/ruleConverters/typedef.ts:
```diff
@@ -0,0 +1,53 @@
+        ...(tslintRule.ruleArguments.some(
+            (item) => item === "call-signature" || item === "arrow-call-signature",
+        ) && {
+            notices: ["Options 'call-signature' and 'arrow-call-signature' are ignored"],
```
Cool, I'll add that in too

hyperupcall commented:
Thank you for your feedback! :smiley:  I will push the improved coverage and address your replies by this Friday (I have some Exams in the meantime)!

JoshuaKGoldberg commented:
Good luck on the exams! ❤️

hyperupcall committed:
style: Condense repetitive code

This also uses `Set` to improve time complexity of searching to `O(n)`
time

hyperupcall committed:
test: Fix to use `@jest/globals`

hyperupcall committed:
test: Add case for all argument conversions

hyperupcall committed:
feat: Complete `typedef` support

This adds full support to `explicit-function-return-type` and
`explicit-module-boundary-type`

hyperupcall commented:
I've implemented the changes, including use of `Set()`, `Object.entries()`, and got back to 100% codecov :smile: 

Also, both `explicit-function-return-type` and `explicit-module-boundary-types` are now used in the conversion process. Since I'm not super familiar with TSLint, I manually tested each one to be 100% sure conversion was the same. When this conversion is done, I added a message to to `notices` since ESLint seems to only lint all functions (not classifying between arrow functions and non-arrow functions as TSLint does with `arrow-call-signature` and `call-signature`. I tried my best to make it succinct, lemmie know if you have any improvements

JoshuaKGoldberg reviewed (approved):
Super, thanks so much @hyperupcall! 🙌

JoshuaKGoldberg commented:
This should be available in tslint-to-eslint-config@2.12.0.
